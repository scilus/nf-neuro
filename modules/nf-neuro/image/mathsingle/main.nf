process IMAGE_MATHSINGLE {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_2.0.2.sif':
        'scilus/scilus:2.0.2' }"

    input:
        tuple val(meta), path(image)

    output:
        tuple val(meta), path("*.nii.gz")        , emit: image
        path "versions.yml"                      , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    assert task.ext.operation in ['lower_threshold', 'upper_threshold',
    'lower_threshold_eq', 'upper_threshold_eq', 'lower_threshold_otsu',
    'upper_threshold_otsu', 'lower_clip', 'upper_clip', 'absolute_value',
    'round', 'ceil', 'floor', 'normalize_sum', 'normalize_max',
    'log_10', 'log_e', 'convert', 'invert', 'dilation', 'erosion',
    'closing', 'opening', 'blur'] : "Operation ${task.ext.operation} not \
    supported. Supported operations are: \
    'lower_threshold', 'upper_threshold', 'lower_threshold_eq', \
    'upper_threshold_eq', 'lower_threshold_otsu', 'upper_threshold_otsu', \
    'lower_clip', 'upper_clip', 'absolute_value', 'round', 'ceil', \
    'floor', 'normalize_sum', 'normalize_max', 'log_10', 'log_e', \
    'convert', 'invert', 'dilation', 'erosion', 'closing', 'opening', \
    'blur'"

    def prefix = task.ext.prefix ?: "${meta.id}"
    def suffix = task.ext.suffix ?: "output"
    def value = task.ext.value ?: ""
    def data_type = task.ext.data_type ?: "float32"
    def exclude_background = task.ext.exclude_background ? "--exclude_background" : ""
    """
    scil_volume_math.py ${task.ext.operation} $image $value \
        ${prefix}__${suffix}.nii.gz --data_type $data_type $exclude_background

        cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def suffix = task.ext.suffix ?: "output"
    """
    scil_volume_math.py -h
    touch ${prefix}_${suffix}.nii.gz


    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """
}
