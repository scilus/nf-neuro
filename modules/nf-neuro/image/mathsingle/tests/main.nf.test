nextflow_process {

    name "Test Process IMAGE_MATHSINGLE"
    script "../main.nf"
    process "IMAGE_MATHSINGLE"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfcore"
    tag "image"
    tag "image/mathsingle"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "antsbet.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("image - lower_threshold") {
        config "./nextflow.config"

        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_brain_probability_map.nii.gz")
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("image - stub") {
        options "-stub-run"
        config "./nextflow.config"

        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_brain_probability_map.nii.gz")
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.versions).match() }
            )
        }
    }
}


