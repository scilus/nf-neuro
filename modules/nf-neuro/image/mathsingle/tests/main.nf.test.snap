{
    "image - stub": {
        "content": [
            [
                "versions.yml:md5,8f977a45a021f6093638e92e1ba4efc1"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-05-29T12:12:28.292893548"
    },
    "image - lower_threshold": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__thr.nii.gz:md5,cb81e82131721a035c6c3a897113dc77"
                    ]
                ],
                "1": [
                    "versions.yml:md5,8f977a45a021f6093638e92e1ba4efc1"
                ],
                "image": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__thr.nii.gz:md5,cb81e82131721a035c6c3a897113dc77"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,8f977a45a021f6093638e92e1ba4efc1"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-05-29T12:12:24.10037522"
    }
}